src/crimson/net/Messenger.cc

```diff
@@ -6,16 +6,28 @@
 
 namespace ceph::net {
 
+template <typename T, typename... Args>
+seastar::future<T*>create_sharded_msgr(Args... args) {
```

Comment:
why?

Reply:
messenger shard stop is called before osd::stop, so when do osd::shutdown->heartbeat::shutdown (front_ msgr and back_msgr)are freed before it. cause call virtual function error. they are coupled,  but what I did seems not complete decouple them, I think still has problem.

Reply:
in this fun, I call messenger's wait function wait for messenger shutdown.

Reply:
why we need create messenger shard within OSD shard?  Messenger sharded_obj is not used outside of create_shard.

Reply:
the intention is to make the OSD self-contained and to initialize the facilities where they are used. i don't think it is related to if `sharded_obj` is used in OSD or not.

Reply:
i just realized that i misunderstood you. https://github.com/ceph/ceph/projects/2#card-18842255 is created to track this issue.